File: packages/support-assemblyscript/assembly/service.ts
import { readAsString, readFromMemory, writeToMemory } from './memory';
import * as internal from './milliservices_v1'

@final export class request {
  static metadata(key: string): string {
    const ptr = internal.get_metadata(writeToMemory(String.UTF8.encode(key)));
    return readAsString(ptr);
  }
}

@final export class response {
  static metadata(key: string, value: string): void {
    const key_ptr = writeToMemory(String.UTF8.encode(key));
    const value_ptr = writeToMemory(String.UTF8.encode(value));
    internal.set_response_metadata(key_ptr, value_ptr);
  }

  static sendData(data: ArrayBuffer): void {
    internal.send_response(writeToMemory(data));
  }

  static sendString(data: string): void {
    response.sendData(String.UTF8.encode(data));
  }
}

@final export class ServiceCall {
  private id: u32;
  public response: ServiceResponse;

  constructor(name: string) {
    const key_ptr = writeToMemory(String.UTF8.encode(name));
    const id = internal.service_new_request(key_ptr);
    this.id = id;
    this.response = new ServiceResponse(id);
  }

  setMetadata(key: string, value: string): void {
    const key_ptr = writeToMemory(String.UTF8.encode(key));
    const value_ptr = writeToMemory(String.UTF8.encode(value));
    internal.service_set_metadata(this.id, key_ptr, value_ptr);
  }

  setData(data: ArrayBuffer): void {
    internal.service_write_data(this.id, writeToMemory(data));
  }

  execute(): void {
    internal.service_execute(this.id);
  }
}

@final class ServiceResponse {
  private id: u32;

  constructor(id: u32) {
    this.id = id
  }

  metadata(key: string): string {
    const key_ptr = writeToMemory(String.UTF8.encode(key));
    const val_ptr = internal.service_get_response_metadata(this.id, key_ptr);
    return readAsString(val_ptr);
  }

  data(): ArrayBuffer {
    const ptr = internal.service_get_response(this.id);
    return readFromMemory(ptr);
  }
}

File: packages/support-rust/src/service.rs
use crate::internal;
use crate::memory;

pub struct ServiceRequest(u32);

impl ServiceRequest {
  pub fn new(key: &str) -> Self {
    unsafe {
      let id = internal::service_new_request(memory::write_to_memory(key.into()));
      Self(id)
    }
  }

  pub fn set_body(&self, data: Vec<u8>) {
    unsafe { internal::service_write_data(self.0, memory::write_to_memory(data)) }
  }

  pub fn set_metadata(&self, key: &str, value: &str) {
    unsafe {
      internal::service_set_metadata(
        self.0,
        memory::write_to_memory(key.into()),
        memory::write_to_memory(value.into()),
      )
    }
  }

  pub fn execute(&self) -> ServiceResponse {
    unsafe {
      internal::service_execute(self.0);
    }
    ServiceResponse::new(self.0)
  }
}

pub struct ServiceResponse(u32);

impl ServiceResponse {
  pub fn new(id: u32) -> Self {
    Self(id)
  }

  pub fn data(&self) -> Vec<u8> {
    unsafe { memory::read_from_memory(internal::service_get_response(self.0)) }
  }

  pub fn metadata(&self, key: &str) -> String {
    unsafe {
      let key_ptr = memory::write_to_memory(key.into());
      memory::read_as_string(internal::service_get_response_metadata(self.0, key_ptr))
        .unwrap_or("".to_string())
    }
  }
}
